### PR TITLE
[DAGCombiner] In mergeTruncStore, make sure we aren't storing shifted in bits.

### LINKED ISSUE #90936: miscompile related to coalescing stores in AArch64 SDAG backend
here's a function:
```llvm
define void @f(i16 %0, ptr %1) {
  %3 = trunc i16 %0 to i8
  %4 = trunc i16 %0 to i14
  %new0 = lshr i14 %4, 8
  store i8 %3, ptr %1, align 1
  %5 = getelementptr i8, ptr %1, i64 1
  %6 = trunc i14 %new0 to i8
  store i8 %6, ptr %5, align 1
  ret void
}
```

during codegen, the two 8-bit stores are getting merged:
```
_f:    
	strh	w0, [x1]
	ret
```

but the backend seems to miss that the trunc changed the value, it dropped those 2 high bits. 
I also have versions of this that improperly merge 16-bit stores and 32-bit stores.

we can demonstrate using this driver:
```c
#include <stdio.h>

void f(unsigned short, unsigned char *);

int main(void) {
  unsigned char mem[8];

  for (int i=0; i<8; ++i)
    mem[i] = 0xee;

  f(0xFFFF, mem);

  for (int i=0; i<8; ++i)
    printf("%x\n", mem[i]);
}
```

```
Johns-MacBook-Pro:tmp regehr$ ~/llvm-project/for-alive/bin/llc foo.ll
Johns-MacBook-Pro:tmp regehr$ clang driver.c foo.s && ./a.out 
ff
ff
ee
ee
ee
ee
ee
ee
Johns-MacBook-Pro:tmp regehr$ 
```

in contrast, global isel properly notices that those high bits were chopped off:
```
Johns-MacBook-Pro:tmp regehr$ ~/llvm-project/for-alive/bin/llc foo.ll -global-isel
Johns-MacBook-Pro:tmp regehr$ clang driver.c foo.s && ./a.out 
ff
3f
ee
ee
ee
ee
ee
ee
```

cc @hatsunespica

### PR DESCRIPTION
**Pull request by topperc**
When looking through a right shift, we need to make sure that all of
the bits we are using from the shift come from the shift input and
not the sign or zero bits that are shifted in.
    
Fixes #90936.

### CODE REVIEW

File: llvm/lib/Support/RISCVISAUtils.cpp [hunk 1]
```diff
@@ -35,7 +36,7 @@ enum RankFlags {
 // Get the rank for single-letter extension, lower value meaning higher
 // priority.
 static unsigned singleLetterExtensionRank(char Ext) {
-  assert(Ext >= 'a' && Ext <= 'z');
+  assert(isLower(Ext));
```
Comment (arsenm) [inline llvm/lib/Support/RISCVISAUtils.cpp:39]: Unrelated change

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 2]
```diff
@@ -485,6 +485,11 @@ RISCVISAInfo::parseNormalizedArchString(StringRef Arch) {
     if (MajorVersionStr.getAsInteger(10, MajorVersion))
       return createStringError(errc::invalid_argument,
                                "failed to parse major version number");
+
+    if (ExtName[0] == 'z' && (ExtName.size() == 1 || isDigit(ExtName[1])))
+      return createStringError(errc::invalid_argument,
+                               "'z' must be followed by a letter");
```
Comment (arsenm) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:491]: Unrelated change

File: llvm/test/CodeGen/AArch64/pr90936.ll [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+; CHECK-NEXT:    ubfx w8, w0, #8, #6
+; CHECK-NEXT:    strb w0, [x1]
+; CHECK-NEXT:    strb w8, [x1, #1]
+; CHECK-NEXT:    ret
+  %3 = trunc i16 %0 to i8
+  %4 = trunc i16 %0 to i14
+  %new0 = lshr i14 %4, 8
+ store i8 %3, ptr %1, align 1
+  %5 = getelementptr i8, ptr %1, i64 1
+ %6 = trunc i14 %new0 to i8
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/pr90936.ll:16]: Indentation weird and use named values

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by efriedma-quic** [comment]
Note that we actually could merge the stores here, if we cared... just need to make sure we mask the value before we store it.
